yarl/__init__.py

```diff
@@ -912,6 +912,8 @@ def with_fragment(self, fragment):
             fragment = ""
         elif not isinstance(fragment, str):
             raise TypeError("Invalid fragment type")
+        if self.fragment == fragment:
```

Comment:
```suggestion
        if self.raw_fragment == fragment:
```
self.fragment *decodes* raw_fragment string which adds own overhead